server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java

```diff
@@ -391,39 +390,59 @@ private static InternalAggregations aggregate(
         int resultSetSize,
         AggregationReduceContext reduceContext
     ) {
-        interface ReleasableIterator extends Iterator<InternalAggregations>, Releasable {}
-        try (var aggsIter = new ReleasableIterator() {
-
-            private Releasable toRelease;
-
-            @Override
-            public void close() {
-                Releasables.close(toRelease);
+        if (resultSetSize == 1) {
+            if (partialResults.hasNext()) {
+                return InternalAggregations.reduce(partialResults.next(), reduceContext);
             }
-
-            @Override
-            public boolean hasNext() {
-                return toConsume.hasNext();
+            try (var delayable = toConsume.next().consumeAggs()) {
+                return InternalAggregations.reduce(delayable.expand(), reduceContext);
             }
-
-            @Override
-            public InternalAggregations next() {
-                var res = toConsume.next().consumeAggs();
-                Releasables.close(toRelease);
-                toRelease = res;
-                return res.expand();
+        }
+        try {
+            // general case
+            if (partialResults.hasNext()) {
+                return consumeAggResults(partialResults, toConsume, createReducer(resultSetSize, reduceContext, partialResults.next()));
             }
-        }) {
-            return InternalAggregations.topLevelReduce(
-                partialResults.hasNext() ? Iterators.concat(partialResults, aggsIter) : aggsIter,
-                resultSetSize,
-                reduceContext
-            );
+            AggregatorsReducer reducer;
+            try (var delayable = toConsume.next().consumeAggs()) {
+                reducer = createReducer(resultSetSize, reduceContext, delayable.expand());
+            }
+            return consumeAggResults(partialResults, toConsume, reducer);
         } finally {
             toConsume.forEachRemaining(QuerySearchResult::releaseAggs);
         }
     }
 
+    private static AggregatorsReducer createReducer(int resultSetSize, AggregationReduceContext reduceContext, InternalAggregations first) {
```

Comment:
The follow-up from this would be to create the reducer up-front and in the size of either the final resultset (if no partial reduces come into play) or the size of the next partial reduce and consume aggregations directly into the aggregator as results come in.
I chose to do this in 2 steps as making this performance while receiving results concurrently will require forcing all results onto a single thread which is little code but non-trivial to reason about :)
This change gets us close enough to make it little code though :)